### PR TITLE
Create a proper state enum in LedgerConfigurationSubscriptionFromIndex. [KVL-1046]

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
I think this clarifies the potential states. I also added some missing test cases.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/configuration/LedgerConfigurationSubscriptionFromIndexSpec.scala [hunk 1]
```diff
@@ -115,6 +116,101 @@ final class LedgerConfigurationSubscriptionFromIndexSpec
...
+        scheduler = scheduler,
+        materializer = materializer,
+        servicesExecutionContext = system.dispatcher,
+      )
+
+      subscriptionBuilder
+        .subscription(configurationLoadTimeout)
+        .use { currentLedgerConfiguration =>
+          Delayed.by(100.millis)(()).map { _ =>
+            currentLedgerConfiguration.latestConfiguration() should be(None)
```
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/configuration/LedgerConfigurationSubscriptionFromIndexSpec.scala:149]:
```suggestion
            currentLedgerConfiguration.latestConfiguration should be(None)
```

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/LedgerConfigurationSubscriptionFromIndex.scala [hunk 2]
```diff
@@ -130,7 +136,8 @@ private[apiserver] final class LedgerConfigurationSubscriptionFromIndex(
 
     override def ready: Future[Unit] = readyPromise.future
 
-    override def latestConfiguration(): Option[Configuration] = latestConfigurationState.get._2
+    override def latestConfiguration(): Option[Configuration] =
```
Comment (fabiotudone-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/LedgerConfigurationSubscriptionFromIndex.scala:139]:
```suggestion
    override def latestConfiguration: Option[Configuration] =
```

Or was it intentional?

**Review by fabiotudone-da — APPROVED** [review]

File: ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/configuration/LedgerConfigurationSubscriptionFromIndexSpec.scala (hunk 1 above)
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/test/suite/scala/platform/apiserver/configuration/LedgerConfigurationSubscriptionFromIndexSpec.scala:149]: Nope, this is definitely a method that depends on side effects, and therefore should not be confused with a stateless property.

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/LedgerConfigurationSubscriptionFromIndex.scala (hunk 2 above)
Comment (SamirTalwar-DA) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/LedgerConfigurationSubscriptionFromIndex.scala:139]: Yup, very intentional. I don't like omitting parentheses for referentially-opaque functions.

**Pull request merged**